cpp/src/parquet/arrow/reader.cc

```diff
@@ -462,7 +463,8 @@ class LeafReader : public ColumnReaderImpl {
         input_(std::move(input)),
         descr_(input_->descr()) {
     record_reader_ = RecordReader::Make(
```

Comment:
Should we need to check the `field_` is a "large" type if `use_large_binary_variants`?

Reply:
I don't think so, if the user is asking to use large_variants, shouldn't we just use it?

Reply:
Hm. That's an error on user side, right? He'll get an exception as expected, it's actually a matter of where the exception should be thrown.

Maybe it does make sense to throw an exception earlier, what do you think?

Reply:
Yes I think this patch already cast schema to "Large" if user open a file with this `use_large_binary_variants` option. But in RecordReader there're no any checkings, I guess explicit cast can make the code more clear. But maybe not having a check, and just add comment for `use_large_binary_variants` force using "Large" is also ok

Reply:
Yes, in fact reading from parquet is a bit dirty and hacking. Some user just uses api like only `ColumnWriter` and `ColumnReader` to handle low-level api, only know the record count and binary as "raw records"

Some users uses `parquet::arrow` reader, this will assemble "arrow message" and assemble "raw records" to arrow format.

Maybe we should add test like:

```c++
  void CheckReadValues(int start, int end) {
    auto binary_reader = dynamic_cast<BinaryRecordReader*>(record_reader_.get());
    ASSERT_NE(binary_reader, nullptr);
    // Chunks are reset after this call.
    ::arrow::ArrayVector array_vector = binary_reader->GetBuilderChunks();
    ASSERT_EQ(array_vector.size(), 1);
    auto binary_array =
        dynamic_cast<::arrow::FixedSizeBinaryArray*>(array_vector[0].get());

    ASSERT_NE(binary_array, nullptr);
    ASSERT_EQ(binary_array->length(), record_reader_->values_written());
    if (read_dense_for_nullable()) {
      ASSERT_EQ(binary_array->null_count(), 0);
      ASSERT_EQ(record_reader_->null_count(), 0);
    } else {
      ASSERT_EQ(binary_array->null_count(), record_reader_->null_count());
    }

    std::vector<std::string_view> expected = expected_values(start, end);
    for (size_t i = 0; i < expected.size(); ++i) {
      if (def_levels_[i + start] == 0) {
        ASSERT_EQ(!read_dense_for_nullable(), binary_array->IsNull(i));
      } else {
        ASSERT_EQ(expected[i].compare(binary_array->GetView(i)), 0);
        ASSERT_FALSE(binary_array->IsNull(i));
      }
    }
  }
```

And ensure that it returns the "Large" type.

Reply:
Qq: Is there some sort of documentation that illustrates how all of these Encoders/Decoders/Dictionaries/FileREaders/RecordReaders, arrow and parquet, interact with each other? I checked on https://github.com/apache/arrow/tree/main/docs/source/developers/cpp, but couldn't find it.

It's getting very hard for me to reason about some of the questions you raise like this last comment. I either need to take a step back and investigate how the lib is architectured and how components interact with each other or ask for more help from you.

Sorry for that.

@wgtmac @mapleFU

Reply:
I have tried to connect a few dots and read good portion of the code, here's my current understanding of the architecture. Will probably dig more into it tomorrow.

`FileReader` is the broader arrow interface as far as I understand. It allows the user to read entire columns, tables and also get some readers.

`FileReaderImpl` implements such functionality. One of its functionalities is `GetFieldReader`, which in my naive assumption is used by all of its other api's to get the typed reader. The typed reader will resolve either to a `LeafReader` or a special struct reader that contains `LeafReaders`, that decision is based on whether the parquet/arrow type is special or not.

This FieldReader is probably responsible for parsing parquet specific stuff, will get into it later. `FieldReader` is an implementation of `ColumnReaderImpl`.

`ColumnReaderImpl` implements `ColumnReader`, which according to the docs is just a stream reader. It probably contains states and reads n bytes. The `ByteArrayRecordReader` can be of dictiopnary reader and non dictionary reader. Let me focus on non dict reader for now. It of course inherits / extends `TypedRecordReader`, a class that implements methods like `ReadBatch`. ReadBatch depends on `TypedDecoder` to read the values, that is resolved to `MakeTypedDecoder` response. In this case, `PlainByteArrayDecoder`. Finally, `PlainByteArrayDecoder` is the class that reads the values with the help of `ArrowBinaryHelperBase`. `PlainByteArrayDecoder` explicitly uses 32bit integers, which I assume it's because this won't change for Large / non large since parquet supports only 32bit strings.

I believe `ArrowbinaryHelperBase` is used to store those values in an `arrow` format, which means that's already no longer parquet. It then proceeds to use 64 bit / large variants according to traits, aside from dictionary accumulator;

If it uses the dense binary read path `PlainByteArrayDecoderBase::DecodeArrow` (whatever that means), it'll use LargeBinaryBuilder with 64bit offsets. I have sanity checked it and it looks ok. If it uses the dictionary read path `PlainByteArrayDecoderBase::DecodeArrow`, it'll use the `Dictionary32Builder`. In that case, the interface available is of course 32 bit. Which also seems to be ok because of the previous discussion: parquet does not allow a string that big and we are not expecting a dictionary to contain that many distinct strings.

Reply:
Yes, your understanding is correct. The documentation is not that friendly to someone new to parquet-cpp. :(